File: RemoveComments.c
#include <stdio.h>

int main(void) {
    int c;

    goto ESTADO_CODIGO;

    // ESTADO_CODIGO:
    // - Emite chars normales
    // - Detecta '/', '"' o '\''
    ESTADO_CODIGO:
        c = getchar();
        if (c == EOF) goto FIN;
        if (c == '/') goto ESTADO_BARRA; 
        if (c == '"') {
            putchar(c);
            goto ESTADO_STRING;
        }
        if (c == '/') {
            putchar(c);
            goto ESTADO_CARACTER;
        }
        putchar(c);
        goto ESTADO_CODIGO;
        
    // ESTADO_BARRA:
    // - Tras leer '/'
    // - Determina inicio de comentario
    ESTADO_BARRA:
        c = getchar();
        if (c == EOF) {
            putchar('/');
            goto FIN;
        };
        if (c == '/') goto ESTADO_COM_LINEA;
        if (c == '*') goto ESTADO_COM_BLOQUE;
        putchar('/');
        putchar(c);
        goto ESTADO_CODIGO;
    
    // ESTADO_COM_LINEA:
    // - Descarta hasta encontrar '\n'
    // - Emite '\n' y vuelve a ESTADO_CODIGO
    ESTADO_COM_LINEA:
        c = getchar();
        if (c == EOF) goto FIN;
        if (c == '\n') {
            putchar(c);
            goto ESTADO_CODIGO;
        }
        goto ESTADO_COM_LINEA;

    // ESTADO_COM_BLOQUE:
    // - Descarta hasta encontrar '*'
    ESTADO_COM_BLOQUE:
        c = getchar();
        if (c == EOF) goto FIN;
        if (c == '*') goto ESTADO_ASTERISCO;
        goto ESTADO_COM_BLOQUE;

    // ESTADO_ASTERISCO:
    // - Tras ver '*'
    // - Si aparece '/' entonces se cierra el bloque
    // - Si aparece '*' entonces se sigue en asterisco
    // - Otro, vuelve a bloque
    ESTADO_ASTERISCO:
        c = getchar();
        if (c == EOF) goto FIN;
        if (c == '/') goto ESTADO_CODIGO;
        if (c == '*') goto ESTADO_ASTERISCO;
        goto ESTADO_COM_BLOQUE;

    // ESTADO_STRING:
    // - Literal entre ""
    // - Maneja '\\' para escapes
    ESTADO_STRING:
        c = getchar();
        if (c == EOF) goto FIN;
        putchar(c);
        if (c == '\\') goto ESTADO_ESCAPE;
        if (c == '"') goto ESTADO_CODIGO;
        goto ESTADO_STRING;

    // ESTADO_CARACTER:
    // - Literal entre ''
    // - Igual que CADENA
    ESTADO_CARACTER:
        c = getchar();
        if (c == EOF) goto FIN;
        putchar(c);
        if (c == '\\') goto ESTADO_ESCAPE;
        if (c == '\'') goto ESTADO_CODIGO;
        goto ESTADO_CARACTER;

    // ESTADO_ESCAPE:
    // - Emite el caracter escapado
    // - Vuelve a STRING o CARACTER
    ESTADO_ESCAPE:
        c = getchar();
        if (c == EOF) goto FIN;
        putchar(c);
        goto ESTADO_STRING;
    
    FIN:
        return 0;
}

File: implementaciones/RemoveComments.hs
import System.IO (isEOF)

data Estado = 
    Codigo      -- Estado principal. Copia caracteres normales
  | Barra       -- Detecta '/' y espera ver si es comentario
  | ComLinea    -- Comentario de línea "//". Descarta hasta encontrar '\n'
  | ComBloque   -- Comentario de bloque "/* ... */". Descarta contenido
  | Asterisco   -- Subestado dentro de comentario de bloque, mira '*', posible cierre
  | String      -- Dentro de un literal de texto "", se copia todo
  | Caracter    -- Dentro de un literal de carácter ''
  | Escape      -- Secuencia de escape en literal (\n, \", etc.)
  deriving (Eq)

main :: IO ()
main = process Codigo

-- process: procesa caracter a caracter segun el estado
process :: Estado -> IO ()
process estado = do
  eof <- isEOF
  if eof then return () else do
    c <- getChar
    case (estado, c) of

      -- ESTADO CODIGO: copia todo salvo '/', '"', '\''
      (Codigo, '/') -> process Barra -- Posible inicio de comentario
      (Codigo, '"') -> putChar '"' >> process String -- Literal de string
      (Codigo, '\'') -> putChar '\'' >> process Caracter -- Literal de carácter
      (Codigo, char) -> putChar char >> process Codigo -- Otro: emite y sigue

      -- ESTADO BARRA: decide si es //, /* o solo '/'
      (Barra, '/') -> process ComLinea -- Comentario de línea
      (Barra, '*') -> process ComBloque -- Comentario de bloque
      (Barra, char) -> putChar '/' >> putChar char >> process Codigo -- No era comentario: emite '/' y el carácter

      -- ESTADO COM_LINEA: descarta todo hasta '\n'
      (ComLinea, '\n') -> putChar '\n' >> process Codigo
      (ComLinea, _) -> process ComLinea -- Sigue descartando

      -- ESTADO COM_BLOQUE: descarta todo hasta ver '*'
      (ComBloque, '*') -> process Asterisco -- Observa '*', posible cierre
      (ComBloque, _) -> process ComBloque

      -- ESTADO ASTERISCO: confirma si se cierra con '/'
      (Asterisco, '/') -> process Codigo -- Fin del comentario de bloque
      (Asterisco, '*') -> process Asterisco -- Sigue en posible cierre
      (Asterisco, _)   -> process ComBloque -- No era cierre, vuelve a ComBloque

      -- ESTADO STRING: copia todo hasta que termine el literal
      (String, '\\') -> putChar '\\' >> process Escape -- Secuencia de escape
      (String, '"') -> putChar '"' >> process Codigo -- Cierra literal
      (String, char) -> putChar char >> process String -- Copia carácter

      -- ESTADO CARACTER: similar a STRING pero para comillas simples
      (Caracter, '\\') -> putChar '\\' >> process Escape
      (Caracter, '\'') -> putChar '\'' >> process Codigo
      (Caracter, char) -> putChar char >> process Caracter

      -- ESTADO ESCAPE: emite carácter y vuelve al literal correspondiente
      (Escape, char) -> putChar char >> process String